bindings/wasm/gltfIO.js

```diff
@@ -0,0 +1,208 @@
+}
+
+export function readMesh(mesh, attributes, materials) {
+  if (attributes.length < 1 || attributes[0] !== 'POSITION')
+    throw new Error('First attribute must be "POSITION".');
+
+  const primitives = mesh.listPrimitives();
+  if (primitives.length === 0) {
+    return {};
+  }
+
+  const numProp = attributes.map((def) => attributeDefs[def].components)
+                      .reduce((a, b) => a + b);
+  const position = primitives[0].getAttribute('POSITION');
+  const numVert = position.getCount()
+  const vertProperties = new Float32Array(numProp * numVert);
+  let offset = 0;
+  for (const attribute of attributes) {
+    const accessor = primitives[0].getAttribute(attribute);
+    const array = accessor.getArray();
+    const size = accessor.getElementSize();
+    for (let i = 0; i < numVert; ++i) {
+      for (let j = 0; j < size; ++j) {
+        vertProperties[numProp * i + offset + j] = array[i * size + j];
+      }
+    }
+    offset += size;
+  }
+
+  const triVertArray = [];
+  const runIndexArray = [0];
+  for (const primitive of primitives) {
+    if (primitive.getAttribute('POSITION') === position) {
+      triVertArray.push(...primitive.getIndices().getArray());
+      runIndexArray.push(triVertArray.length);
+      materials.push(primitive.getMaterial());
+    } else {
+      console.log('primitives do not share accessors!');
+    }
+  }
+  const triVerts = new Uint32Array(triVertArray);
+  const runIndex = new Uint32Array(runIndexArray);
+
+  const manifoldPrimitive = mesh.getExtension('EXT_manifold');
+  const mergeTriVert =
+      manifoldPrimitive ? manifoldPrimitive.getMergeIndices().getArray() : [];
+  const mergeTo =
+      manifoldPrimitive ? manifoldPrimitive.getMergeValues().getArray() : [];
+  const vert2merge = new Map();
+  for (const [i, idx] of mergeTriVert.entries()) {
+    vert2merge.set(triVerts[idx], mergeTo[i]);
+  }
+  const mergeFromVert = [];
+  const mergeToVert = [];
+  for (const [from, to] of vert2merge.entries()) {
+    mergeFromVert.push(from);
+    mergeToVert.push(to);
+  }
+
+  return {
+    numProp,
+    triVerts,
+    vertProperties,
+    runIndex,
+    mergeFromVert,
+    mergeToVert
+  };
+}
+
+export function writeMesh(doc, manifoldMesh, attributes, materials) {
+  if (doc.getRoot().listBuffers().length === 0) {
+    doc.createBuffer();
+  }
+  const buffer = doc.getRoot().listBuffers()[0];
+  const manifoldExtension = doc.createExtension(EXTManifold);
+
+  const indices = doc.createAccessor('indices')
+                      .setBuffer(buffer)
+                      .setType(Accessor.Type.SCALAR)
+                      .setArray(manifoldMesh.triVerts);
+
+  const mesh = doc.createMesh();
+  const numPrimitive = manifoldMesh.runIndex.length - 1;
+  for (let run = 0; run < numPrimitive; ++run) {
+    const primitive = doc.createPrimitive().setIndices(indices);
+    const material = materials[run];
+    if (material) {
+      primitive.setMaterial(material);
```

Comment:
@donmccurdy I'm loving your library! However, I'm getting `Error: Cannot connect disconnected graphs.` at this line. I can't quite tell from your docs how I should go about moving a `Material` (along with the textures it references) from a file I read to a file I'm writing. Is there a reason it can't attach them to the requested doc automatically? I was looking for something like `Root.attach()`, but came up empty.

Reply:
There isn't currently a feature that transfers a resource on its own or with its dependencies, though that's an interesting idea. Currently the options would be:

(a) Merge documents, then clean up any unwanted resources. The prune() transform can help with getting rid of unused resources.

```js
documentA.merge(documentB);
documentA.getRoot().listScenes()[1].dispose();

// ... assign materials as needed ...

await documentA.transform(prune());
```

(b) Create a new material in the target document, then deep copy the original.

```js
document.createMaterial().copy(material, (ref) => ref.clone());
```

Reply:
/cc

- https://github.com/donmccurdy/glTF-Transform/issues/924

Reply:
Thanks, that'll unblock me. Looking forward to the new feature though 😄

Reply:
I'm still not sure what's going on here sorry, but I did publish an uncompressed version (no other relevant changes, just not minified) as v4.0.0-alpha.0 if you want to test from that:

```
npm install --save @gltf-transform/core@next @gltf-transform/extensions@next @gltf-transform/functions@next
```

Reply:
Strange, when I update my import map to `https://cdn.jsdelivr.net/npm/@gltf-transform/functions@v4.0.0-alpha.0/+esm` I still get a minified version (maybe something jsdelivr is doing?). When I point directly to your file `https://cdn.jsdelivr.net/npm/@gltf-transform/functions@v4.0.0-alpha.0/dist/functions.modern.js` it doesn't seem to work as a module, since I get `caught TypeError: Failed to resolve module specifier "property-graph". Relative references must start with either "/", "./", or "../".`

Since I can't debug it locally, I'll work on getting a raw.githack working so you can see my repro.

Reply:
Hm, yeah it sounds like jsdelivr must be rewriting the file to resolve imports, and also minifying... that does complicate things. I could try running this PR locally if you could share steps for reproducing the issue here.

Reply:
Perfect, that fixed it! Actually `@3.x` had the same problem, but `@3.2.1` works fine. Curios if that will break again the next time a new version is released? Or was that just something weird about v3.1.2? 

I suppose vitejs is an option, but the point of these examples is to be simple. It's kind of nice if all anyone needs is an import map. One question though: why do I need `/+esm` when your default file appears to be an ES6 module already? It'd be nice to not have to rely on those CDN transformations.

Reply:
Perhaps a piece of the puzzle: when I try `https://cdn.jsdelivr.net/npm/@gltf-transform/core@3.2.1` I get `Failed to load module script: Expected a JavaScript module script but the server responded with a MIME type of "application/node". Strict MIME type checking is enforced for module scripts per HTML spec.` - I wonder if that's a bug with jsDelivr or if the extensions need to be `.mjs` or something?

Reply:
> Curious if that will break again the next time a new version is released?

Unfortunately, yes. The CDN is dumber than a bundler, it doesn't know what version(s) are already on the page, so it just resolves the newest compatible version of /core.

> Expected a JavaScript module script but the server responded with a MIME type of "application/node" ...

It looks like JSDelivr will serve the (`.cjs`) "main" entry by default, and the `/+esm`  tag is required to request something else. Maybe this new esm.run thing by JSDeliver would be the way to go? It isn't something I've used before, though:

https://www.jsdelivr.com/esm

Reply:
Ah, interesting. Well this works, and I think should be reasonably future-proof since I'm guessing you at least restrict by major version:
```
    "imports": {
      "@gltf-transform/core": "https://cdn.jsdelivr.net/npm/@gltf-transform/core@3/+esm",
      "@gltf-transform/functions": "https://cdn.jsdelivr.net/npm/@gltf-transform/functions@3/+esm"
    }
```

Reply:
Ugh, I guess I should just switch to TypeScript. Probably a good idea anyway, especially for manifoldCAD.org. I just need to bite the bullet.

Reply:
I do restrict to same major version, not sure why `@3/+esm` doesn't work. ☹️

Reply:
Ah, I get it now - the `/+esm` thing is adding exact versions to your import statements and allowing them to automatically resolve. The reason `/dist/core.modern.js` isn't working is that I need to add all of your imports to my import map. I'm beginning to realize why bundling and tree-shaking is so necessary.